main.rb

```diff
@@ -1,17 +1,160 @@
+require 'byebug'
+
 module Enumerable
-    def my_each
-        for i in 0...self.size
-            yield(self[i])
-        end
-    self
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def my_each
+    i = 0
+    if block_given?
+      while i < length
+        yield(self[i])
+        i += 1
+      end
+      return self
+    end
+    to_enum
+  end
+
+  def my_each_with_index
+    i = 0
+    if block_given?
+      while i <= length - 1
+        yield(self[i], i)
+        i += 1
+      end
+      return self
+    end
+    to_enum
+  end
+
+  def my_select
+    arr = []
+    return to_enum unless block_given?
+
+    my_each { |x| arr << x if yield(x) == true }
+    arr
+  end
+
+  def my_all?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == false }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == false }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == false }
+    elsif arg.nil? == false
+      my_each { |x| return false if x != arg }
+    else
+      my_each { |x| return false unless x }
+    end
+    true
+  end
+
+  def my_any?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return true if prc.call(x) }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return true if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return true if x.is_a?(arg) }
+    elsif arg.nil? == false
+      my_each { |x| return true if x == arg }
+    else
+      my_each { |x| return true if x }
+    end
+    false
+  end
+
+  def my_none?(arg = nil, &prc)
+    if block_given?
+      my_each { |x| return false if prc.call(x) == true }
+    elsif arg.is_a?(Regexp)
+      my_each { |x| return false if arg.match?(x.to_s) == true }
+    elsif arg.is_a?(Class)
+      my_each { |x| return false if x.is_a?(arg) == true }
```

Comment:
- my_none?

- [ ] when no block or argument is given returns true only if none of the collection members is true. Check out [this documentation](https://apidock.com/ruby/Enumerable/none%3F)
`[false, nil].my_none?` should returns true
![Screenshot from 2020-04-22 08-52-45](https://user-images.githubusercontent.com/30977845/79978681-ac4dab00-8476-11ea-837d-00c6663eaa70.png)